src/data/markdown/translated-guides/en/07 Misc/05 k6 Extensions.md

```diff
@@ -1,4 +1,197 @@
+`.\k6` on Windows) to avoid using any other `k6` binary in your `PATH`.
+
+
+### Writing a new extension
+
+A good starting point for using xk6 and writing your own extension is the [xk6
+introductory article](https://k6.io/blog/extending-k6-with-xk6).
+<!-- TODO: How much of this article do we want to reproduce here? -->
+
+You'll first need to decide the type of extension you wish to create. Choose a
+JavaScript extension if you want to extend the JS functionality of your script, or an
+Output extension if you want to process the metrics emitted by k6 in some way.
+
+A few things to keep in mind when writing a new extension:
+<!-- TODO: Better structure this list? -->
+
+- JS extensions are registered using
+  [`modules.Register()`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/js/modules#Register)
+  and Output extensions using
+  [`output.RegisterExtension()`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/output#RegisterExtension),
+  which should be called in the `init()` function of the module.
+  In either case you'll have to specify a unique name that preferably doesn't clash
+  with any existing extension. JS extensions also must begin with `k6/x/` so as to
+  differentiate them from built-in JS modules.
+
+- JS extensions expose Go methods which can optionally receive a
+  [`context.Context`](https://golang.org/pkg/context/#Context) instance as the first argument.
+  This context is used throughout the k6 codebase and contains embedded objects
+  which can be extracted to access the
+  [`goja.Runtime`](https://pkg.go.dev/github.com/dop251/goja#Runtime) for a
+  particular VU, and to get more information about the test in progress.
+  Take a look at
+  [`common.GetRuntime()`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/js/common#GetRuntime),
+  [`lib.GetState()`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/lib#GetState) and
+  [`lib.GetExecutionState()`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/lib#GetExecutionState).
+
+- Output extensions must implement the
+  [`output.Output`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/output#Output) interface.
+  They should be particularly weary of performance, since they can receive a large
+  amount of metrics to process, which can easily degrade the performance of the test
+  if the extension is inefficient.
+  As such you should ensure that `AddMetricSamples()` doesn't block for a long time,
+  and that metrics are flushed periodically to avoid memory leaks.
+
+  Since this is common functionality that most outputs should have, we've provided
+  a couple of helper structs: [`output.SampleBuffer`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/output#SampleBuffer)
+  and [`output.PeriodicFlusher`](https://pkg.go.dev/go.k6.io/k6@v0.33.0/output#PeriodicFlusher)
+  that output implementations can use which handles bufferring and periodic
+  flushing in a thread-safe way.
+  For usage examples see the [`statsd` output](https://github.com/k6io/k6/blob/v0.33.0/output/statsd/output.go#L55).
```

Comment:
Can you add an example of an extension instead of the k6 codebase?

Reply:
Hhmm we already have a lot of external examples and I thought it would be good to link to an internal module for once, to showcase that extensions and internal modules are near identical in implementation.

Besides, I'm not sure there are any extensions that actually use this yet, since it hasn't been documented ;)